File: web/src/api/savedMedicines.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) OwnPulse Contributors

import { api } from "./client";

export interface SavedMedicine {
  id: string;
  substance: string;
  dose: number | null;
  unit: string | null;
  route: string | null;
  sort_order: number;
  created_at: string;
}

export interface CreateSavedMedicine {
  substance: string;
  dose?: number;
  unit?: string;
  route?: string;
}

export const savedMedicinesApi = {
  list: () => api.get<SavedMedicine[]>("/api/v1/saved-medicines"),
  create: (data: CreateSavedMedicine) => api.post<SavedMedicine>("/api/v1/saved-medicines", data),
  update: (id: string, data: Partial<CreateSavedMedicine>) =>
    api.put<SavedMedicine>(`/api/v1/saved-medicines/${id}`, data),
  remove: (id: string) => api.delete<void>(`/api/v1/saved-medicines/${id}`),
};

File: backend/api/src/routes/saved_medicines.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) OwnPulse Contributors

use axum::Json;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use uuid::Uuid;

use crate::AppState;
use crate::auth::extractor::AuthUser;
use crate::db::saved_medicines as db;
use crate::error::ApiError;
use crate::models::saved_medicine::{CreateSavedMedicine, SavedMedicineRow, UpdateSavedMedicine};

/// GET /saved-medicines
pub async fn list(
    State(state): State<AppState>,
    AuthUser { id: user_id, .. }: AuthUser,
) -> Result<Json<Vec<SavedMedicineRow>>, ApiError> {
    let rows = db::list(&state.pool, user_id).await?;
    Ok(Json(rows))
}

/// POST /saved-medicines
pub async fn create(
    State(state): State<AppState>,
    AuthUser { id: user_id, .. }: AuthUser,
    Json(body): Json<CreateSavedMedicine>,
) -> Result<(StatusCode, Json<SavedMedicineRow>), ApiError> {
    if body.substance.trim().is_empty() {
        return Err(ApiError::BadRequest(
            "substance must not be empty".to_string(),
        ));
    }
    let row = db::insert(&state.pool, user_id, &body).await?;
    Ok((StatusCode::CREATED, Json(row)))
}

/// PUT /saved-medicines/:id
pub async fn update(
    State(state): State<AppState>,
    AuthUser { id: user_id, .. }: AuthUser,
    Path(id): Path<Uuid>,
    Json(body): Json<UpdateSavedMedicine>,
) -> Result<Json<SavedMedicineRow>, ApiError> {
    let row = db::update(&state.pool, user_id, id, &body)
        .await?
        .ok_or(ApiError::NotFound)?;
    Ok(Json(row))
}

/// DELETE /saved-medicines/:id
pub async fn delete(
    State(state): State<AppState>,
    AuthUser { id: user_id, .. }: AuthUser,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, ApiError> {
    let deleted = db::delete(&state.pool, user_id, id).await?;
    if deleted {
        Ok(StatusCode::NO_CONTENT)
    } else {
        Err(ApiError::NotFound)
    }
}
